lib/signalfx/signalflow/websocket.rb

```diff
@@ -0,0 +1,303 @@
+        # that gains very little in terms of efficiecy given how quick auth
+        # should be.
+        start_time = Time.now
+        while !@authenticated
+          # The socket will be closed by the server if auth isn't successful
+          # within 5 seconds so no point in waiting longer
+          if Time.now - start_time > 5 || @close_reason
+            raise "Could not authenticate to SignalFlow WebSocket: #{@close_reason}"
+          end
+          sleep 0.1
+        end
+      end
+
+      @ws.send(msg)
+    end
+  end
+  private :send_msg
+
+  def on_close(msg)
+    @close_reason = "(#{msg.code}, #{msg.data})"
+    @chan_callbacks.keys.each do |channel_name|
+      invoke_callback_for_channel({ :event => "CONNECTION_CLOSED" }, channel_name)
+    end
+
+    reinit
+  end
+
+  def on_message(m)
+    begin
+      return if m.type == :ping
+      if m.type == :close
+        on_close(m)
+        return
+      end
+
+      message_received(m.data, m.type == :text)
+    rescue Exception => e
+      puts "Error processing SignalFlow message: #{e.backtrace.first}: #{e.message} (#{e.class})"
+    end
+  end
+
+  def on_open
+    @ws.send({
+      :type => "authenticate",
+      :token => @api_token,
+    }.to_json)
+  end
+
+  # Start up a new WS client in its own thread that runs an EventMachine
+  # reactor.
+  def startup_client
+    this = self
+    WebSocket::Client::Simple.connect("#{@stream_endpoint}/v2/signalflow/connect") do |ws|
+      @ws = ws
+      ws.on :error do |e|
+        puts "ERROR #{e.inspect}"
+      end
+
+      ws.on :close do |e|
+        this.on_close(e)
+      end
+
+      ws.on :message do |m|
+        this.on_message(m)
+      end
+
+      ws.on :open do
+        this.on_open
+      end
+    end
+  end
+  private :startup_client
+
+  def invoke_callback_for_channel(msg, channel_name)
+    chan = @chan_callbacks[channel_name]
+
+    raise "Callback for channel #{channel_name} is missing!" unless chan
+
+    if chan == DETACHED
+      return
+    else
+      chan.inject_message(msg)
+    end
+  end
+  private :invoke_callback_for_channel
+
+  def message_received(raw_msg, is_text)
+    msg = add_parsed_timestamp!(parse_message(raw_msg, is_text))
+
+    if msg[:type] == "authenticated"
+      @authenticated = true
+      return
+    end
+
+    if msg[:channel]
+      invoke_callback_for_channel(msg, msg[:channel])
+    else
+      # Ignore keep-alives
+      if msg[:event] == "KEEP_ALIVE"
+        return
+      else
+        raise "Unknown SignalFlow message: #{msg}"
+      end
+    end
+  end
+  private :message_received
+
+  def parse_message(raw_msg, is_text)
+    if is_text
+      JSON.parse(raw_msg, {:symbolize_names => true})
+    else
+      BinaryMessageParser.parse(raw_msg)
+    end
+  end
+  private :parse_message
+
+  def add_parsed_timestamp!(msg)
+    if msg.has_key?(:timestampMs)
+      msg[:timestamp] = Time.at(msg[:timestampMs]/1000)
```

Comment:
Same as earlier, you probably want to `/ 1000.0` to maintain millisecond precision